src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1601,14 +1617,21 @@ internal bool DeclaresTheObjectClass
                         this.GetSymbolsWithNameCore(WellKnownMemberNames.EntryPointMethodName, SymbolFilter.Member, cancellationToken));
 
                     // Global code is the entry point, ignore all other Mains.
-                    var scriptClass = this.ScriptClass;
-                    if (scriptClass is object)
+                    if (scriptClass is object || simpleProgramEntryPointSymbol is object)
                     {
                         foreach (var main in entryPointCandidates)
                         {
                             diagnostics.Add(ErrorCode.WRN_MainIgnored, main.Locations.First(), main);
                         }
-                        return scriptClass.GetScriptEntryPoint();
+
+                        if (scriptClass is object)
+                        {
+                            return scriptClass.GetScriptEntryPoint();
+                        }
+
+                        RoslynDebug.Assert(simpleProgramEntryPointSymbol is object);
+                        entryPointCandidates.Clear();
+                        entryPointCandidates.Add(simpleProgramEntryPointSymbol);
```

Comment:
> It looks like the remaining pieces of this method are validating that the entry point is acceptable, but since we're synthesizing it I presume we already know that. Why not just return the symbol immediately?

This is the candidate and I do not mind do for it what we would do for any other candidate. I believe this is more robust approach than the suggested one because we don't need to worry if any changes below made in the future invalidate the assumption that it is safe to return.

---
In reply to: [382854192](https://github.com/dotnet/roslyn/pull/41706#discussion_r382854192) [](ancestors = 382854192)